physics/module_mp_thompson.F90

```diff
@@ -3742,19 +3703,21 @@ subroutine mp_thompson (qv1d, qc1d, qi1d, qr1d, qs1d, qg1d, ni1d, &
          qv1d(k) = MAX(1.E-10, qv1d(k) + qvten(k)*DT)
          qc1d(k) = qc1d(k) + qcten(k)*DT
          nc1d(k) = MAX(2./rho(k), MIN(nc1d(k) + ncten(k)*DT, Nt_c_max))
-         nwfa1d(k) = MAX(11.1E6/rho(k), MIN(9999.E6/rho(k),             &
+         nwfa1d(k) = MAX(11.1E6, MIN(9999.E6,                           &
```

Comment:
@climbfuji I am not sure about the change on line 3745. I think the units of nwfa1d are modified, and later in the code nwfa1 is multiplied by rho. I already wrote about my concern on Greg's PR, but he seemed not to agree.

Reply:
In lines 1788-1789 in the new/modified code, we have:
```
         nwfa(k) = MAX(11.1E6, MIN(9999.E6, nwfa1d(k)*rho(k)))
         nifa(k) = MAX(naIN1*0.01, MIN(9999.E6, nifa1d(k)*rho(k)))
```
In line 2990, we have:
```
         nwfa(k) = MAX(11.1E6, (nwfa1d(k) + nwfaten(k)*DT)*rho(k))
```
and then, in lines 3706-3709:
```
         nwfa1d(k) = MAX(11.1E6, MIN(9999.E6,                           &
                       (nwfa1d(k)+nwfaten(k)*DT)))
         nifa1d(k) = MAX(naIN1*0.01, MIN(9999.E6,                       &
                       (nifa1d(k)+nifaten(k)*DT)))
```
@tanyasmirnova and gthompsnWRF, the previous lines are assigning values of nwfa1d to nwfa by multiplying it with rho. But lines 3706-3709 are only updating nwfa1d itself, using the same lower/upper limits as for nwfa, and I think therefore @tanyasmirnova is correct that we need to divide by rho here?

But, isn't this an unnecessary operation anyway? After lines 3706-3709, these values of nwfa1d/ninfa1d are not used anymore, and since they are local variables there values are "forgotten"?

Reply:
I am hopeful we can put this issue to rest.  The variable nwfa(k) is local inside the main code subroutine and represents aerosols in number per cubic meter.  The variable nwfa1d is a vertical column passed in via gt_driver (where it went from 3D to 1D) and absolutely represents number per kg of air.  So line#3706 is **critical** to eliminate the air density or else the number that goes back to the model for advection can be massive when rho is tiny such as up at 100-200km altitude.  That is the bug fix for some models with ultra-high top.

I've decided to throw in the multiplication by rho for the Min/Max settings just to ensure no tendency gets created at the far ends of the range in the case that the model drifts off beyond one of the ranges.  It shouldn't really be much chance of that, but it's best I don't take that chance.

But, let me emphasize how important line#3706 is that the values must not be divided by rho.

Reply:
Thank you, @gthompsnWRF, all this sounds reasonable to me. You know this code best after all ... @tanyasmirnova are you also satisfied with this explanation?

Reply:
@DomHeinzeller   I just pushed another commit to my branch for #567 dealing with this item, hopefully for final time.

Reply:
Yes, I am fine with the current code. 
Actually, I revisited our work with HRRR physics in the global FIM, and found that I did the same change myself in September 2017:
@@ -3547,10 +3553,14 @@
          qv1d(k) = MAX(1.E-10, qv1d(k) + qvten(k)*DT)
          qc1d(k) = qc1d(k) + qcten(k)*DT
          nc1d(k) = MAX(2./rho(k), nc1d(k) + ncten(k)*DT)
-         nwfa1d(k) = MAX(11.1E6/rho(k), MIN(9999.E6/rho(k),             &
-                       (nwfa1d(k)+nwfaten(k)*DT/rho(k))))
-         nifa1d(k) = MAX(naIN1*0.01/rho(k), MIN(9999.E6/rho(k),                &
-                       (nifa1d(k)+nifaten(k)*DT)/rho(k)))
+!         nwfa1d(k) = MAX(11.1E6/rho(k), MIN(9999.E6/rho(k),             &
+!tgs                       (nwfa1d(k)+nwfaten(k)*DT/rho(k))))
+         **nwfa1d(k) = MAX(11.1E6, MIN(9999.E6,             &
+                       (nwfa1d(k)+nwfaten(k)*DT)))**
+!         nifa1d(k) = MAX(naIN1*0.01/rho(k), MIN(9999.E6/rho(k),                &
+!tgs                       (nifa1d(k)+nifaten(k)*DT)/rho(k)))
+         **nifa1d(k) = MAX(naIN1*0.01, MIN(9999.E6,                &
+                       (nifa1d(k)+nifaten(k)*DT)))**

Greg, sorry, that I haven't checked my FIM code earlier. I agree, let's put this issue to rest now.

Reply:
Yay.